File: Metodos Iterativos de Optimizacion Numerica de Funciones de Varias Variables/Gradiente Conjugado No Lineal/Gradiente Conjugado_No_Lineal.py
import numpy as np
from sympy import *
from math import *
import random

def gradient(f,vars):
    output=[]
    for var in vars:
        output.append(sympify(diff(f,var)))
    return output
def gradientEval(gradient,vars,values):
    output=[]
    for ind in range(len(vars)):
        output.append(gradient[ind].subs(zip(vars,values)))
    return output
def norm(gradient):
    suma=0
    for component in gradient:
        suma+=component**2
    return sqrt(suma)
def gcnl(f,variables,iniciales,tol,iterMax):
    var_temp=variables.split(" ")
    val_temp=iniciales.split(" ")
    vars=[]
    xk=[]
    for ind in range(len(var_temp)):
        vars.append(var_temp[ind])
        xk.append(float(val_temp[ind]))

    f=sympify(f)
    grdf=gradient(f,vars)
    gk=gradientEval(grdf,vars,xk)
    dk=[]
    for valor in gk:dk.append(-valor)
    k,error=0,0
    while(k<iterMax):
        delta=random.random()
        ak=0.5
        aux=[]
        for ind in range(len(xk)):aux.append(xk[ind]+ak*dk[ind])
        izq=f.subs(zip(vars,aux))-f.subs(zip(vars,xk))
        der=0
        for ind in range(len(gk)):der+=delta*ak*gk[ind]*dk[ind]
        while(izq>der):
            ak=ak/2
            aux = []
            for ind in range(len(xk)): aux.append(xk[ind] + ak * dk[ind])
            izq = f.subs(zip(vars, aux)) - f.subs(zip(vars, xk))
            der = 0
            for ind in range(len(gk)): der += delta * ak * gk[ind] * dk[ind]
        xkp1=[]
        for ind in range(len(xk)):xkp1.append(xk[ind]+ak*dk[ind])
        xk=xkp1
        error=norm(gradientEval(grdf,vars,xk))
        if(abs(error)<=tol):
            return (xk,abs(error),k)
        gkp1=gradientEval(grdf,vars,xk)
        beta=(((norm(gkp1))**2)/((norm(gk))**2))
        gk=gkp1
        dkp1=[]
        for ind in range(len(dk)):dkp1.append(-gk[ind]+beta*dk[ind])
        dk=dkp1
        k+=1
    return (xk, abs(error), k)


f="(x-2)^4+(x-2*y)^2"
variables="x y"
iniciales="3 0"#Componentes del vector inicial
tol=1e-5
iterMax=2500
valores=gcnl(f,variables,iniciales,tol,iterMax)
print("El punto aproximado es "+str(valores[0])+" con un error de "+str(valores[1])+" en "+str(valores[2])+" iteraciones")

File: Metodos Iterativos de Optimizacion Numerica de Funciones de Varias Variables/Gradiente Conjugado No Lineal/Gradiente_Conjugado_No_Lineal.m
function Gradiente_Conjugado_No_Lineal
  clc;clear;
  pkg load symbolic;
  f="(x+9)^2+y^2+5";
  variables="x y";
  iniciales="3 0"; %Componentes del vector inicial
  tol=1e-5;
  iterMax=2500;
  [xk,error,k]=gcnl(f,variables,iniciales,tol,iterMax);
  display(["El punto aproximado por el metodo es (",num2str(xk(1)),",",num2str(xk(2)) ,") con un error de ",num2str(error)," en ",num2str(k)," iteraciones"]);
end

function[xk,error,k]=gcnl(f,variables,iniciales,tol,iterMax)
  warning('off', 'all');
  eval(["syms " variables ";"])
  eval(["vars=[" variables "];"]);
  eval(["xk=[" iniciales "]';"]);%Debe ser traspuesto para poder operarlo con gk y dk
  f=sym(f);
  grdf=gradient(f,vars);
  grs=sym(grdf);
  gk=double(subs(grs,vars,xk));
  dk=-gk;
  k=0;error=0;
  
  while k<iterMax
    delta=rand(1);
    ak=0.5;
    aux=xk+ak*dk;
    izq=double(subs(f,vars,aux));
    der=delta*ak*gk'*dk;
    while izq>der
      ak=ak/2;
      aux=xk+ak*dk;
      izq=double(subs(f,vars,aux)-subs(f,vars,xk));
      der=delta*ak*gk'*dk;
    endwhile
    xk=xk+ak*dk;
    error=norm(double(subs(grs,vars,xk)))
    if abs(error)<=tol
      error=abs(error);
      return
    endif
    gkp1=double(subs(grs,vars,xk));
    beta=norm(gkp1)^2/norm(gk)^2;
    gk=gkp1;
    dk=-gk+beta*dk;
    k=k+1;
  endwhile
  error=abs(error);
  return;
end